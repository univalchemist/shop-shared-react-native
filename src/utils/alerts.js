import { Platform, ActionSheetIOS, Alert } from 'react-native';

function showConfirmation({
  title,
  message,
  actionTitle,
  cancelTitle,
  destructive,
  onAction,
}) {
  if (Platform.OS === 'ios') {
    ActionSheetIOS.showActionSheetWithOptions(
      {
        title,
        message,
        options: [cancelTitle, actionTitle],
        destructiveButtonIndex: destructive ? 1 : null,
        cancelButtonIndex: 0,
      },
      buttonIndex => {
        if (buttonIndex === 1) {
          onAction();
        }
      },
    );
  } else {
    Alert.alert(
      title,
      message,
      [
        { text: cancelTitle, style: 'negative', onPress: () => {} },
        {
          text: actionTitle,
          onPress: onAction,
          style: 'positive',
        },
      ],
      { cancelable: false },
    );
  }
}

function debounceAlert({ subject, message, buttons, options, delay = 500 }) {
  setTimeout(() => {
    Alert.alert(subject, message, buttons, options);
  }, delay);
}

export { showConfirmation, debounceAlert };
