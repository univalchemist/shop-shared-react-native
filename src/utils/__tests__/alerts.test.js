import { Alert, Platform, ActionSheetIOS } from 'react-native';
import { showConfirmation, debounceAlert } from '../alerts';

jest.mock('react-native/Libraries/Alert/Alert', () => ({ alert: jest.fn() }));
jest.mock('react-native/Libraries/ActionSheetIOS/ActionSheetIOS', () => ({
  showActionSheetWithOptions: jest.fn(),
}));

describe('showConfirmation', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('should show ios specific actionsheet', () => {
    Platform.OS = 'ios';

    showConfirmation({
      title: undefined,
      message: undefined,
      destructive: true,
      actionTitle: 'test action',
      cancelTitle: 'cancel',
    });

    expect(ActionSheetIOS.showActionSheetWithOptions).toHaveBeenCalled();
  });
  it('should call onAction on ios specific actionsheet action', () => {
    Platform.OS = 'ios';
    const onActionSpy = jest.fn();

    showConfirmation({
      title: undefined,
      message: undefined,
      destructive: true,
      actionTitle: 'test action',
      cancelTitle: 'cancel',
      onAction: onActionSpy,
    });
    ActionSheetIOS.showActionSheetWithOptions.mock.calls[0][1](1);

    expect(onActionSpy).toHaveBeenCalled();
    expect(
      ActionSheetIOS.showActionSheetWithOptions.mock.calls[0],
    ).toMatchSnapshot();
  });
  it('should call onAction on ios specific actionsheet action with destructive false', () => {
    Platform.OS = 'ios';
    const onActionSpy = jest.fn();

    showConfirmation({
      title: undefined,
      message: undefined,
      destructive: false,
      actionTitle: 'test action',
      cancelTitle: 'cancel',
      onAction: onActionSpy,
    });
    ActionSheetIOS.showActionSheetWithOptions.mock.calls[0][1](1);

    expect(onActionSpy).toHaveBeenCalled();
    expect(
      ActionSheetIOS.showActionSheetWithOptions.mock.calls[0],
    ).toMatchSnapshot();
  });

  it('should show android specific alert', () => {
    Platform.OS = 'android';

    showConfirmation({
      title: 'testing',
      message: 'test message',
      destructive: true,
      actionTitle: 'test action',
      cancelTitle: 'cancel',
    });

    expect(Alert.alert).toHaveBeenCalled();
  });

  it('should call onAction on android specific Alert action', () => {
    Platform.OS = 'android';
    const onActionSpy = jest.fn();

    showConfirmation({
      title: 'testing',
      message: 'test message',
      actionTitle: 'test action',
      cancelTitle: 'cancel',
      onAction: onActionSpy,
    });
    Alert.alert.mock.calls[0][2][1].onPress();

    expect(onActionSpy).toHaveBeenCalled();
  });
});

describe('debounceAlert', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('should show ios specific actionsheet after default 500ms', () => {
    debounceAlert({
      title: 'testing',
      message: 'test message',
    });

    setTimeout(() => {
      expect(Alert.alert).toHaveBeenCalled();
    }, 500);
  });

  it('should show ios specific actionsheet after delay', () => {
    const delay = 1000;
    debounceAlert({
      title: 'testing',
      message: 'test message',
      delay,
    });

    setTimeout(() => {
      expect(Alert.alert).toHaveBeenCalled();
    }, delay);
  });
});
