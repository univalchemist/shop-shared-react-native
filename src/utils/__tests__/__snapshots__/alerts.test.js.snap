// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showConfirmation should call onAction on ios specific actionsheet action 1`] = `
Array [
  Object {
    "cancelButtonIndex": 0,
    "destructiveButtonIndex": 1,
    "message": undefined,
    "options": Array [
      "cancel",
      "test action",
    ],
    "title": undefined,
  },
  [Function],
]
`;

exports[`showConfirmation should call onAction on ios specific actionsheet action with destructive false 1`] = `
Array [
  Object {
    "cancelButtonIndex": 0,
    "destructiveButtonIndex": null,
    "message": undefined,
    "options": Array [
      "cancel",
      "test action",
    ],
    "title": undefined,
  },
  [Function],
]
`;
